Migrate to Prosopo Procaptcha

forgetso opened the pull request:
Fixes #337

PR to migrate rococo faucet from reCAPTCHA to Prosopo Procaptcha, which runs on the rococo test network.

##  Overview

When using Procaptcha, the flow of data in the faucet **client** app is as follows :

### Frontend Flow

1. The user enters the page.
2. The user selects their network and how many tokens they require.
3. The user clicks the Procaptcha check box.
4. A Polkadot `user_account` is generated using fingerprinting as random entropy
5. The `user_account` and the faucet's `site_key` are passed to the *Procaptcha Contract* in rococo, and details of a random Captcha Provider are returned.
6. The rococo network is contacted, querying the *Procaptcha Contract*, and a Captcha Provider is returned
7. The website asks the Captcha Provider for a challenge, passing the `site_key`, `user_account`, `block_number`, and other data related to the Captcha Provider.
8. The Captcha Provider checks it was randomly selected by the *Procaptcha Contract*, by submitting the same read performed in step 5. using historical block data (e.g. `queryAt(block_number)`)
9. Assuming step 8. is successful, the Captcha Provider returns a challenge
10. The user completes the challenge, and passes the solution to the Captcha Provider. If the solution is correct, the checkbox becomes ticked, a challenge payload is added to the form, and the form submission button becomes enabled
11. The user presses the form submission button and the form data + captcha payload are sent to the faucet backend

#### Additional Info

##### Page Reloads

We noticed that the reCAPTCHA implementation would ask the user for a new captcha on every page load. We have recreated this functionality using procaptcha by clearing any local storage on load of the `CaptchaV2.svelte` component. This prevents a locally stored Captcha Provider from being contacted on page reload, and passing the `user_account` due to their historical records stored on the Captcha Provider.

##### Max Verified Time

The maximum amount of time a captcha solution is valid for is 60s. This can be changed using the env variable `PROCAPTCHA_MAX_VERIFIED_TIME`, which takes a millisecond amount.

### Backend Flow

We'll begin the backend flow from where the frontend flow left off.

1. The backend receives the form data from the frontend.
2. The backend takes the Procaptcha payload and passes it to a TypeScript package called `@prosopo/server`.
3. This package performs the following checks
  a. Check if the provider was actually chosen at `block_number`
  b. If a `provider_url` is passed with the payload, the Captcha Provider is contacted to verify the user. Otherwise, read the `user_account` score from the contract directly after checking the time since the last correct captcha is less than the `maxVerifiedTime` (`60s`, mentioned previously)
4. After 3a. or 3b. the user is either verified via a Captcha Provider or the Procaptcha Contract and will be permitted or denied their funds.

### Testing

#### Client

Our client bundle continues to reach out to the rococo network for a random provider and the provider sends a *real* captcha. Solving this is clearly out of scope for UI tests so the provider is patched by injecting our mock provider (https://mockprovider.prosopo.io) into local storage. The mock provider always returns `{ "success": true }` for the Alice account and the test `site_key=5C4hrfjw9DjXZTzV3MwzrrAr9P1MJhSrvWGWqi1eSuyUpnhM`, which is the all zero address.

#### E2E

As we are using a contract to return a provider, we implemented an additional contracts parachain in the E2E tests. This requires us to do the following:

1. deploy the Procaptcha Contract
2. register a Captcha Provider in the contract (We register our mock provider - https://mockprovider.prosopo.io)
3. store a Captcha Provider dataset in the contract
4. register an app (`site_key`) in the contract

Then when a random provider is selected, it is always the mock provider. Again, passing a `user_account` of Alice's address and the zero address for `site_key` returns `{ "success": true }`. Obviously, the downside here is the time taken for transactions in 12s parachains in zombienet. We have tried to use the substrate-contracts-node, which has instant seal, alongside the zombienet configuration but didn't succeed.

We would appreciate any feedback on the approach so far!

Bullrich reviewed (commented):
Hi @forgetso! Thank you for the PR.

First of all, thank you a lot for adding tests! That's amazing!

> More detail to follow...

Does this mean that the PR is still draft? 

---

@mutantcornholio could you please look into the faucet (not client) code? I would appreciate a second set of eyes

forgetso commented:
> Hi @forgetso! Thank you for the PR.
> 
> First of all, thank you a lot for adding tests! That's amazing!

No, problem. It wouldn't be a proper integration without them although it was a mission to build them! 

> > More detail to follow...
> 
> Does this mean that the PR is still draft?

It's 99% of the way there, I plan to write a proper overview of everything that's changed and why because there are a lot of changes. The only reason I'm not saying 100% is because there is an occasionally failing test and I can't work out why it's happening.

I'll try to get the summary done now so that you guys can review with additional context.

forgetso commented on client/package.json:
```diff
@@ -6,7 +6,9 @@
     "preview": "vite preview",
-    "test": "playwright test",
+    "test:recaptcha": "playwright test --config playwright.config.recaptcha.ts",
+    "test:procaptcha": "playwright test --config playwright.config.procaptcha.ts",
```
No problem at all, there will be a big explainer for tests in the main description soon.

forgetso commented on client/README.md:
```diff
@@ -10,9 +10,11 @@ Two current options are to [access Matrix and contact a bot](https://wiki.polkad
 
-- `PUBLIC_CAPTCHA_KEY`: The [reCaptcha v2 site key](https://www.google.com/u/0/recaptcha/admin).
+- `PUBLIC_CAPTCHA_PROVIDER`: The captcha provider. Currently `procaptcha` and `recaptcha` are supported. You will then need one of the following site keys:
+    - `PUBLIC_PROSOPO_SITE_KEY`: The [Prosopo site key](https://prosopo.io/) which is `5HUBceb4Du6dvMA9BiwN5VzUrzUsX9Zp7z7nSR2cC1TCv5jg`.
```
Yes, this is generated by us but you can generate your own by registering on our website. I think it would be best if we hop on a call to go through everything next week. I'll reach out to you on Element to arrange.

forgetso commented:
I've added a summary of the changes @Bullrich, hope it helps :slightly_smiling_face:

Bullrich commented on e2e/zombienet.native.toml:
```diff
@@ -21,3 +21,15 @@ id = 1000
+id = 1002
+chain = "contracts-rococo-local"
+
+  [parachains.collator]
+  name = "contracts"
+  ws_port = 9988
+  command = "polkadot-parachain"
+  args = ["-lparachain=debug", "--discover-local"]
```
Minor formating

```suggestion
[parachains.collator]
name = "contracts"
ws_port = 9988
command = "polkadot-parachain"
args = ["-lparachain=debug", "--discover-local"]
```

Bullrich commented on src/dripper/DripRequestHandler.ts:
```diff
@@ -20,22 +21,26 @@ const isParachainValid = (parachain: string): boolean => {
-      return { error: "Captcha validation was unsuccessful" };
+    if (external) {
+      const captchaValidate = await this.captchaService.validate(opts.captchaResponse);
+      console.log("Captcha validate response", JSON.stringify(captchaValidate));
```
Could you use the logger from `src/logger.ts` please?

```suggestion
      logger.debug(`Captcha validate response ${JSON.stringify(captchaValidate)}`);
```

Bullrich commented on src/dripper/polkadot/PolkadotActions.ts:
```diff
@@ -96,7 +96,9 @@ export class PolkadotActions {
   async teleportTokens(dripAmount: bigint, address: string, parachain_id: string): Promise<DripResponse> {
-    logger.info("💸 teleporting tokens");
+    logger.info(
+      `💸 teleporting ${dripAmount} tokens to ${address} from ${this.account ? this.account.address.toString() : ""} to parachain ${parachain_id}`,
```
I don't know if it is possible to simplify it like this

```suggestion
      `💸 teleporting ${dripAmount} tokens to ${address} from ${this.account?.address.toString() ?? ""} to parachain ${parachain_id}`,
```

forgetso replied:
Will give it shot :)

Bullrich reviewed (commented):
Thank you for updating the body of the PR. Everything is very detailed.

I haven't find any problems with the code in the `client` directory, but, I have a couple of questions:

1. I tried to create an account in https://prosopo.io/#signup but I never received the email. Could you please look into this? I would like to be able to manage the key and store it as a secret.

<img width="805" alt="image" src="https://github.com/paritytech/polkadot-testnet-faucet/assets/8524599/cd01623f-9a9d-454d-81be-c2bd2699cc59">

2. I'm unable to test the client in my computer.
   - When loading the project (`yarn dev`) and clicking on the captcha, I'm getting the error `ProviderUrlUsed`. I haven't modified any file, any idea how/what could be causing this?

forgetso commented:
> I tried to create an account in https://prosopo.io/#signup but I never received the email. Could you please look into this? I would like to be able to manage the key and store it as a secret.

Hey @Bullrich, I think the email might have been delayed. Have you now received it? 

> I'm unable to test the client in my computer.
When loading the project (yarn dev) and clicking on the captcha, I'm getting the error ProviderUrlUsed. I haven't modified any file, any idea how/what could be causing this?

The contract that's on rococo is slightly out of date with the one in our sources. We plan to update this very soon. The error you should have received is `DappDoesNotExist` because your local site key is not registered in our live contract on rococo. You can try using the site key that you should hopefully now have received by email instead. Please let me know if this works!

mutantcornholio commented on package.json:
```diff
@@ -90,5 +96,19 @@
+      "unplugged": true
+    }
+  },
+  "resolutions": {
+    "@polkadot/api": "10.11.1",
+    "@polkadot/api-augment": "10.11.1",
+    "@polkadot/keyring": "12.6.1",
+    "@polkadot/util": "12.6.1",
+    "@polkadot/util-crypto": "12.6.1",
+    "@polkadot/wasm-crypto": "7.2.1",
+    "@polkadot/x-randomvalues": "12.6.1"
```
what does this serve? As I understand what `resolutions` do, they are forcing specific versions of sub-dependencies, in the format of `<dependency>/<subdependency>: <subdependency_version>`, but here, there are only top-level dependencies listed

forgetso replied:
When I've worked with `resolutions`, it's been enough to declare them at the top level of the workspace, unlike npm `overrides`. Declaring them once at the top level is meant to affect nested versions according to the [original RFC](https://github.com/yarnpkg/rfcs/blob/master/implemented/0000-selective-versions-resolutions.md#example).

The reason I had to do this is the auto incrementing you get when putting `^12.6.1` for polkadot versions causes the TypeScript types to go out of sync. `ApiPromise` and other things like `WeightV2` suddenly do not match. This is a constant source of bother with PJS which is why we always fix our versions. 

I'm currently upgrading our packages to `12.6.2`, which is why I haven't updated this PR for a bit. I should hopefully get the latest and greatest PJS into this PR next week.

forgetso replied:
PJS versions in this PR will now be latest so I'll remove resolutions.

forgetso replied:
Resolutions removed.

mutantcornholio commented on package.json:
```diff
@@ -27,19 +27,19 @@
+    "'./src/**/*.js'": [
       "prettier --write",
       "eslint --fix"
     ],
-    "./src/**/*.ts": [
+    "'./src/**/*.ts'": [
       "prettier --write",
       "eslint --fix"
     ],
-    "./client/**/*.js": [
+    "'./client/**/*.js'": [
       "cd client && prettier --write",
       "cd client && eslint --fix"
     ],
-    "./client/**/*.ts": [
+    "'./client/**/*.ts'": [
```
could you elaborate on these changes?

forgetso replied:
This prettier command doesn't work on Ubuntu without doing this.

mutantcornholio commented on env.faucet.config.json:
```diff
@@ -78,6 +78,32 @@
+        "default": "procaptcha",
+        "type": "string"
+      },
+      "PROCAPTCHA_MAX_VERIFIED_TIME": {
+        "description": "The maximum time in ms that a user can be verified for",
+        "default": 60000,
+        "type": "number"
+      },
```
I think this should belong in the constants of the code.  
It's not a secret, and I can't believe that we will configure it differently for different environments.

forgetso replied:
This is done btw.

mutantcornholio commented on package.json:
```diff
@@ -90,5 +96,19 @@
     "typescript": "^4.9.3",
     "typescript-transform-paths": "^3.4.6"
+  },
+  "dependenciesMeta": {
+    "buildcheck@0.0.6": {
+      "unplugged": true
+    }
+  },
```
Is this left on purpose?

forgetso replied:
Yes, I was getting [this](https://github.com/mscdex/buildcheck/issues/1#issuecomment-1119903300) error until I put this in the package.json. I can try removing and seeing if it is still an issue as I recently dist upgraded to Ubuntu 22 from 20.

forgetso replied:
To confirm, I removed this and everything still works.

mutantcornholio commented on src/dripper/Procaptcha.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  getValidator() {
+    console.log("Procaptcha secret is ", this.secret);
```
Forgotten `console.log`? :)
Also, a bunch of `console.log`'s down below

forgetso replied:
Yeah, totally. I'll remove all `console.log`'s :)

mutantcornholio commented on src/faucet.e2e.ts:
```diff
@@ -184,8 +207,13 @@ describe("Faucet E2E", () => {
+      procaptchaDetails.testAccount,
+    );
+    const result = await drip(webEndpoint, userAddress, undefined, randomProvider);
+    console.log("result", result);
```
forgotten console.log

mutantcornholio commented on src/test/setupE2E.ts:
```diff
@@ -1,25 +1,29 @@
+  setupProcaptcha
+} from "src/test/setupE2EProcaptcha";
+import {ApiPromise} from "@polkadot/api";
+
```
why did all the formatting change?

forgetso replied:
Not sure, I will revert this.

forgetso replied:
Reverted.

mutantcornholio commented on src/test/setupE2EProcaptcha.ts:
```diff
@@ -0,0 +1,136 @@
+
+        return {contract, contractAddress: contract.address, testAccount: alicePair.address, siteKey: siteKey}
+    } catch (e) {
+        console.error(e)
+        throw new Error (`Failed to setup Procaptcha: ${JSON.stringify(e, null ,4)}`)
```
Rule of thumb: never handle (including logging) and throw errors at the same time. Either one, or another. Otherwise, it leads to double-logging, or to invalid logging (e.g. the error is logged, although it was handled on the upper level)

UPD: do we even need this try-catch and the wrapping at all?
UPD 2: the same point in the methods below

forgetso replied:
Sure, I'll remove the `console.error`.

In terms of the `try` `catch` I thought it added a bit of context to the issue that occurred during setup. Quite happy to remove them and let a higher up `try` `catch` do the work with a more generalised message.

mutantcornholio replied:
I think it's better to remove it. 
`setupProcaptcha` in the stacktrace tells pretty much the same thing about the context.

forgetso replied:
Removed.

mutantcornholio commented on src/test/setupE2EProcaptcha.ts:
```diff
@@ -0,0 +1,136 @@
+        // any account can register an app site key so using Bob will be fine
+        await procaptchaAppRegister(contract, siteKey);
+
+        // get random active provider and block number from response
+        //const randomProvider = await procaptchaGetRandomProvider(contract, siteKey, alicePair.address)
```
Should it be here?

forgetso replied:
No, will remove.

forgetso replied:
Removed.

mutantcornholio commented on src/dripper/Procaptcha.spec.ts:
```diff
@@ -0,0 +1,62 @@
+    });
+  });
+
+  afterAll((done) => {
```
this function doesn't need a callback, you can remove `done` parameter and the call

forgetso replied:
Sure.

forgetso replied:
Removed `done`

mutantcornholio reviewed (changes requested):
Left a whole bunch of questions, but there is an independent big one: I don't think we need two different captcha systems. Let's remove recaptcha completely. If Procapthca won't work for some reason, we can revert the changes, but I can't see a reason in supporting two.

forgetso commented:
> I don't think we need two different captcha systems

Hey @mutantcornholio, I'd be happy to remove the ad tracking reCAPTCHA out of the faucet app. Let's see how the PR goes and take it from there. The option to have 2 captcha systems was implemented due to [previous request](https://github.com/paritytech/polkadot-testnet-faucet/issues/337#issuecomment-1798169357) in the issue. Maybe you guys could discuss and let me know which way you want to go.

Cheers

mutantcornholio commented:
I see. I actually missed that. We had an internal discussion now and our opinions have diverged (and looks like I'm in minority). 
Soo, let's keep recaptcha for now.

Bullrich commented:
Hi @forgetso!

How's the PR coming along?

Has all the requested changes been addressed? 

I know it's troublesome, but could you also resolves the conflicts please? Once this is done we'll jump to give it another review 🙂 

Thanks!

forgetso commented:
Hey @Bullrich, apologies for the radio silence. I actually have a [second PR in progress](https://github.com/prosopo/polkadot-testnet-faucet/pull/4) which will merge into our forked copy. This addresses all of the issues and merges all upstream changes. The most recent upstream changes have caused a bit of difficulty due to an [upstream polkadot-sdk bug](https://github.com/prosopo/polkadot-testnet-faucet/pull/4#issuecomment-2080000256). However, I'm hoping to have an update to the PR for you to review this week.

forgetso commented:
Hey @Bullrich, I think the PR should now pass if you run the tests. I'll keep checking for updates. Cheers

mutantcornholio commented:
@forgetso the unit tests run has failed with 
```
FAIL src/dripper/Procaptcha.spec.ts (12.627 s)
  ● Console
    console.error
      ERROR (app): AxiosError {
        code: 'ERR_BAD_RESPONSE',
        config: {
          adapter: [ 'xhr', 'http' ],
          data: '{"dapp":"5HUBceb4Du6dvMA9BiwN5VzUrzUsX9Zp7z7nSR2cC1TCv5jg","user":"5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"}',
// here go 8k lines of this console.log 😅
      } ⭕ An error occurred when validating captcha
      48 |       return await prosopoServer.isVerified(prosopoOutput, maxVerifiedTime || this.maxVerifiedTime);
      49 |     } catch (e) {
    > 50 |       logger.error(`⭕ An error occurred when validating captcha`, e);
         |              ^
      51 |       return false;
      52 |     }
      53 |   }
      at Logger.log (node_modules/@eng-automation/js/src/logger.ts:122:9)
      at Logger.error (node_modules/@eng-automation/js/src/logger.ts:158:12)
      at Procaptcha.validate (src/dripper/Procaptcha.ts:50:14)
      at Object.<anonymous> (src/dripper/Procaptcha.spec.ts:58:20)
  ● Prosopo Procaptcha › Validates captcha positively
    expect(received).toBeTruthy()
    Received: false
      51 |     const procaptcha = new Procaptcha(PROSOPO_SITE_KEY);
      52 |     const result = await procaptcha.validate(PROCAPTCHA_VALID_CAPTCHA);
    > 53 |     expect(result).toBeTruthy();
         |                    ^
      54 |   });
      55 |
      56 |   it("Validates captcha negatively", async () => {
      at Object.<anonymous> (src/dripper/Procaptcha.spec.ts:53:20)
Test Suites: 1 failed, 5 passed, 6 total
Tests:       1 failed, 50 passed, 51 total

```

and `prettier --check` has failed with this:
```
$ prettier ./src ./client/src ./client/tests --check
Checking formatting...
[warn] src/dripper/DripRequestHandler.ts
[warn] src/server/routes/actions.spec.ts
[warn] client/src/app.css
[warn] client/src/app.html
[warn] Code style issues found in 4 files. Forgot to run Prettier?
```

The error in e2e tests looks papi-related, let's ignore it for now

mutantcornholio commented on src/test/setupE2EProcaptcha.ts:
```diff
@@ -1,136 +1,252 @@
-        //const randomProvider = await procaptchaGetRandomProvider(contract, siteKey, alicePair.address)
+    get siteKey(): string {
+        return this._siteKey
+    }
```
Let's just make these properties public please

forgetso replied:
Done.

mutantcornholio commented on src/test/setupE2EProcaptcha.ts:
```diff
@@ -1,136 +1,252 @@
+        } catch (e) {
+            throw new Error(`Failed to setup Procaptcha: ${JSON.stringify(e, null, 4)}`)
         }
+    }
```
Could you also remove wrapping everything in try-catches from this file as well? 
Callstacks do a better job of showing where the error happened rather than wrapping everything in try-catch, just to throw the error right in the catch block

forgetso replied:
Done.

Bullrich commented on client/src/lib/components/CaptchaV2.svelte:
```diff
@@ -1,9 +1,10 @@
 
   export let captchaKey: string;
-
+  export let captchaProvider: string;
```
Just a tiny change here for type safety:
```suggestion
  export let captchaProvider: CaptchaProvider;
```

forgetso replied:
Done.

Bullrich commented on client/src/lib/components/CaptchaV2.svelte:
```diff
@@ -22,21 +23,43 @@
       dispatch("token", token);
+      // Forces a new captcha on page reload
+      if (captchaProvider === CaptchaProvider.procaptcha) {
+        window.localStorage.removeItem("@prosopo/current_account");
+        window.localStorage.removeItem("@prosopo/provider");
```
Out of curiosity, why are these elements removed?

forgetso replied:
My bad, these items should have been updated to read

```
window.localStorage.removeItem("@prosopo/procaptcha");
```

We have a mechanism in place that stores the following in a hex string in local storage:

```
export const ProsopoLocalStorageSchema = z.object({
    account: z.string().optional(),
    blockNumber: z.number().optional(),
    providerUrl: z.string().optional(),
})
```

If all 3 are present, it means the user has completed a captcha before from `providerUrl`. When the user returns to the page, we contact the provider and ask if the user was previously verified. If they were, we don't force them to do a captcha again. *However*, in your reCAPTCHA implementation the user was forced to complete a captcha on each page load. We mimic this behaviour by removing the local storage.

It's worth noting that the default verified time is set to 60s. So even if there was local storage, it would only be valid for 60s. Happy to change the way this works if you like?

Bullrich commented on client/src/lib/components/CaptchaV2.svelte:
```diff
@@ -52,16 +75,25 @@
 
 <svelte:head>
   {#if componentMounted}
```
I don't know if this line is needed, tho this was before your changes so you can ignore it.

Bullrich commented on client/src/lib/components/NetworkDropdown.svelte:
```diff
@@ -18,7 +18,7 @@
     </div>
   </div>
-  <ul tabindex="-1" class="dropdown-content menu p-2 shadow bg-base-100 rounded-box w-full text-white">
+  <ul tabindex="-1" class="dropdown-content menu p-2 shadow bg-base-100 rounded-box w-full text-white z-10">
```
Good call with the `z-10`. I noticed that error but didn't have time to solve it.

Bullrich commented on client/tests/faucet.ts:
```diff
@@ -1,36 +1,62 @@
 };
 
-const getFormElements = async (page: Page, getCaptcha = false) => {
+type CaptchaProvider = "recaptcha" | "procaptcha";
+
+const getFormElements = async (page: Page, captchaProvider: "recaptcha" | "procaptcha", getCaptcha = false) => {
```
Just so we can use the type declared above

```suggestion
type CaptchaProvider = "recaptcha" | "procaptcha";

const getFormElements = async (page: Page, captchaProvider: CaptchaProvider, getCaptcha = false) => {
```

forgetso replied:
Wow, good point :man_facepalming:

forgetso replied:
Hey @Bullrich do you have any shared code/types between client and the backend? Both backend and frontend need to share this type and a getter for it to be able to cast from .env `string` type to `CaptchaProvider` type.

e.g.

```typescript
// provider comes from .env file so is string type
export const getCaptchaProvider = (provider: string): CaptchaProvider => {
  if (provider === CaptchaProvider.procaptcha) return CaptchaProvider.procaptcha;
  if (provider === CaptchaProvider.recaptcha) return CaptchaProvider.recaptcha;
  throw new Error(`⭕ - Invalid captcha provider: ${provider}`);
};
```

Bullrich replied:
Unfortunately we don't have any.

There are a couple of duplicated clases/objects, but it didn't make sense to take the work to unify everything. 

For example, the `networkData` object exists both in `src/networkData.ts` and `client/src/lib/utils/networkData.ts`.

Feel free to duplicate the code

Bullrich commented on package.json:
```diff
@@ -71,6 +74,7 @@
     "matrix-js-sdk": "^26.1.0",
     "pg": "^8.11.2",
     "prom-client": "^14.2.0",
+    "react": "^18.2.0",
```
Why was react added as a dependency?

forgetso replied:
Our captcha widget renders with react. Our packages import `@prosopo/common` which has `"react-i18next` as a dependency. It's a package for doing translations in react. It depends on react. Unfortunately, when I install with `yarn`, react is not installed in the dependency tree. I had to manually add it here to force installation.

This was a while ago so I'll try removing this, removing node_modules and doing the installation again to confirm this is still the case.

forgetso replied:
Have removed react as a dep

Bullrich commented on package.json:
```diff
@@ -81,6 +85,7 @@
     "@types/express": "^4.17.13",
     "@types/jest": "^29.4.0",
     "@types/node": "^18.16",
+    "@types/react": "^18",
```
Same here with the types for react.

forgetso replied:
Same as above. Will confirm tomorrow if the install is still an issue.

forgetso replied:
Have removed react as a dep and [all is well](https://github.com/prosopo/polkadot-testnet-faucet/pull/8).

Bullrich commented on src/dripper/DripRequestHandler.ts:
```diff
@@ -20,22 +21,26 @@ const isParachainValid = (parachain: string): boolean => {
-      return { error: "Captcha validation was unsuccessful" };
+    if (external) {
+      const captchaValidate = await this.captchaService.validate(opts.captchaResponse);
+      console.log("Captcha validate response", JSON.stringify(captchaValidate));
```
Use the `logger` here located in `src/logger.ts` please

forgetso replied:
Done

Bullrich commented on src/dripper/polkadot/PolkadotActions.ts:
```diff
@@ -1,6 +1,6 @@
 import { Keyring } from "@polkadot/keyring";
 import { KeyringPair } from "@polkadot/keyring/types";
-import { waitReady } from "@polkadot/wasm-crypto";
+import { cryptoWaitReady } from "@polkadot/util-crypto";
```
@mutantcornholio any thougths on this?

mutantcornholio replied:
Not really know how it is implemented, but if it works, fine by me :)

We'll need to migrate it to PAPI anyway

forgetso replied:
We're using PJS instead of PAPI here because there's no clear route to deploy contracts or making calls / txs with PAPI at the moment. Definitely agree with updating this to PAPI in the future.

Bullrich reviewed (changes requested):
I have a couple of notes, but overall it looks good from my side! Good job!

@mutantcornholio can you look thoroughly into the server side as I'm not that knowledgeable in that area? Thanks!

forgetso commented on client/package.json:
```diff
@@ -20,10 +22,13 @@
     "daisyui": "^4.4.4",
     "postcss": "^8.4.31",
+    "prettier": "^3.0.3",
+    "prettier-plugin-svelte": "^3.0.3",
```
I removed these and everything is still working :man_shrugging:

Bullrich commented on src/dripper/DripRequestHandler.ts:
```diff
@@ -20,22 +21,24 @@ const isParachainValid = (parachain: string): boolean => {
-    private actions: PolkadotActions,
-    private recaptcha: Recaptcha,
-  ) {}
+  // eslint-disable-next-line
```
Are you sure this is necessary?

I just tried **on my machine** (which is not source of truth) to remove this line, and I ran:
- `yarn run fix`
- `yarn run lint`

and I didn't have any issues with the linter.

forgetso replied:
Ok, let me try this locally again and report back.

forgetso replied:
Would you mind pasting the output of

```
> yarn list prettier
yarn list v1.22.22
warning Filtering by arguments is deprecated. Please use the pattern option instead.
├─ @prosopo/typechain-polkadot@1.1.15
│  └─ prettier@2.8.8
├─ prettier-plugin-compactify@0.1.6
│  └─ prettier@2.8.3
└─ prettier@3.0.2
Done in 0.88s.

```

and

```
yarn list eslint
yarn list v1.22.22
warning Filtering by arguments is deprecated. Please use the pattern option instead.
├─ @prosopo/typechain-polkadot@1.1.15
│  └─ eslint@8.57.0
└─ eslint@8.47.0
Done in 0.90s.
```

? Thanks

forgetso replied:
No matter what I do I can't get this check to succeed: https://github.com/prosopo/polkadot-testnet-faucet/actions/runs/8924631461/job/24511372506?pr=9

rzadp replied:
> Are you sure this is necessary?
> 
> I just tried **on my machine** (which is not source of truth) to remove this line, and I ran:
> 
>     * `yarn run fix`
> 
>     * `yarn run lint`
> 
> 
> and I didn't have any issues with the linter.

Don't forget about third step:

* `yarn run fix`
* `yarn run lint`
* `yarn run format`

`lint` is `eslint` and `format` is `prettier`. They're conflicting with each other in a weird way.
If you make `format` green, `lint` goes red, and vice versa.

rzadp replied:
As it is likely an already-existing misconfiguration of lint+prettier in the repository that allows this kind linting conflict to happen, I propose we keep the `eslint-disables` and create an issue to tackle it separately, to not block this PR.

Bullrich commented on src/server/routes/actions.spec.ts:
```diff
@@ -35,8 +35,8 @@ jest.mock("../../config", () => {
         (key: string) =>
-          // eslint-disable-next-line security/detect-object-injection
-          ({ NETWORK: "rococo" })[key], // minimal viable config on the initial import
+          // eslint-disable-next-line
+          ({ NETWORK: "rococo" }[key]), // minimal viable config on the initial import
```
Same here, I don't have the issue if I run `fix` and `lint`.

rzadp replied:
Same as https://github.com/paritytech/polkadot-testnet-faucet/pull/372#discussion_r1587665946

Bullrich reviewed (approved):
It looks good from my side!

All my concerns have been addressed (with the exception of those two small lint ones)

forgetso commented on client/src/lib/components/CaptchaV2.svelte:
```diff
@@ -13,7 +14,7 @@
 
   onMount(() => {
-    window.captchaLoaded = () => {
+    window.captchaLoaded = async () => {
```
No, it's not - good spot! Will remove.

forgetso replied:
This is done.

forgetso commented:
@Bullrich any idea when this will actually merge into main? I'd like to ensure we are around to support when it's released.

Bullrich commented:
> @Bullrich any idea when this will actually merge into main? I'd like to ensure we are around to support when it's released.

@mutantcornholio any blockers from your side? Or can we merge?

Bullrich commented:
@forgetso FYI the faucet tests are failing: https://github.com/paritytech/polkadot-testnet-faucet/actions/runs/8986761444/job/24684989457?pr=372

I re-run it 3 times and it is still failing 😢

forgetso commented:
> @forgetso FYI the faucet tests are failing: https://github.com/paritytech/polkadot-testnet-faucet/actions/runs/8986761444/job/24684989457?pr=372
> 
> I re-run it 3 times and it is still failing 😢

Hey, sorry! I had it passing in my PR... strange. Will look into this today.

mordamax commented:
@forgetso fyi there are some conflicts to resolve :(

forgetso commented:
> @forgetso fyi there are some conflicts to resolve :(

Hey @mordamax, thanks for alerting me. I'm out of office until the 27th but will resolve next week.
